### PR TITLE
Add updateMany, updateManyBy methods (#668)

### LINKED ISSUE #668: Batch update (updateMany)
I am building a UI where users update multiple objects at once. I'd like to update them with a single SQL statement.

Currently, I am doing this with raw SQL (simplified example):

```ts
const updates = [
  { id: 1, name: "Alice Updated" },
  { id: 2, name: "Bob Updated" },
  { id: 3, name: "Charlie Updated" },
]

const values = sqlValues(updates, { id: "int", name: "varchar" }) // more on this below

await db.$query`
  UPDATE "user" AS u
  SET name = v.name
  FROM (VALUES ${values}) AS v(id, name)
  WHERE u.id = v.id
`
```

Alternatively, I can use `updateFrom` but this still needs raw SQL for the values and for simpler cases it's even more boilerplate:

```ts
await db.user
  .with("values", q => q.select("id", "name").fromSql`(SELECT * FROM (VALUES ${values}) v(id, name))`)
  .updateFrom("values", q => q.on("values.id", "user.id"))
  .set({
    name: q => q.ref("values.name"),
  })
```

In either case, generating `values` needs fragile boilerplate:

```ts
// `raw` is marked as "@deprecated use `sql` instead", but:
// - `sql` is not even exported, so the deprecation message is misleading;
// - `sqlFn` is exported but generates invalid SQL provided the same input.
import { raw as sql } from "orchid-orm"

/**
 * Build a parametrized sql fragment for VALUES:
 * ($1::type, $2::type), ($3::type, $4::type), ...
 *
 * To be used for UPDATE ... FROM (VALUES ...).
 *
 * @param rows - array of object rows
 * @param fields - map field name → SQL type (e.g. { id: "uuid", title: "varchar" })
 */
export function sqlValues(rows: Record<string, unknown>[], fields: Record<string, string>) {
  const keys = Object.keys(fields)
  const types = Object.values(fields)
  const values: unknown[] = []
  const rowParts: string[] = []
  let i = 0
  for (const row of rows) {
    const colParts: string[] = []
    for (let k = 0; k < keys.length; k++) {
      values.push(row[keys[k]!])
      colParts.push(`$${i++}::${types[k]}`)
    }
    rowParts.push(`(${colParts.join(", ")})`)
  }
  return sql({ raw: rowParts.join(", "), values: values as Record<number, unknown> })
}
```

I don't really like having to provide the type map; I tried to play with `UNION ALL` hoping it would help Postgres to infer types, but to no avail:

```sql
UPDATE "user" AS t
SET name = v.name
FROM (
  SELECT * FROM (
    SELECT id, name FROM "user" WHERE false
    UNION ALL
    VALUES ${values} -- Error: UNION types integer and text cannot be matched
  ) t
) AS v
WHERE t.id = v.id
```

---

This way or another, I'd like this to be supported in the query builder.

The API could look like:

```ts
await db.user.updateMany(
  [
    { id: 1, name: "Alice Updated" },
    { id: 2, name: "Bob Updated" },
    { id: 3, name: "Charlie Updated" },
  ],
  { where: "id" }, // or ["statusId", "kind"] for multi-column
)
```

I anticipate SQL types would come from `UserTable.columns` (as I gathered, that's what `insertMany` uses internally with `insertManyFromValuesAs`).

Perhaps `where` could be more flexible to cover more cases, but honestly I don't feel competent enough to suggest how.

### PR DESCRIPTION
**Pull request by IlyaSemenov**
Add `updateMany`, `updateManyOptional`, `updateManyBy`, `updateManyByOptional` methods (resolves #668).

Following some observations and questions — please review and approve:

## MOST IMPORTANT: Duplicate keys check

- `updateMany` validates input before generating SQL: key columns must be present, key values must be concrete values (not null/undefined, and **not expressions**). User input having duplicate keys is rejected early. Perhaps, for `updateMany*Optional` we could allow expressions and disable duplicate check?
- Duplicate composite keys are compared via `JSON.stringify()`. This is simple and adequate for normal scalar DB keys, but it is still a heuristic rather than a database-level equality model.

## API semantic

- Row shape is determined **from the first row**, and every following row must match exactly.
- `.set()` values cannot “fill in” missing per-row columns.
- Per-row values always win over shared `.set()` values and over hook-generated values for the same column.
- Empty effective set of updated fields is an error. Unlike regular `update`, batch update does not fall back to `SELECT`.
- `updateMany` is intentionally incompatible with `join` and `updateFrom`; both chaining orders are rejected at runtime.

## Implementation decisions

- `InsertQueryDataObjectValues` was removed because it was barely used and similar code already worked with `unknown[][]` directly.

## Technical debt

- The local `ShapePrimaryKeys` type is a simplified version of `ShapeColumnPrimaryKeys`. Should this be unfiied somehow (especially if we allow expression for `updateMany*Optional`)?
- `UpdateManyResult` is a literal copy/paste of `InsertManyResult`. Should they be unified somehow?
- `updateManyBy` does **not** runtime-verify that `keys` match a declared unique constraint. That remains type-level only, consistent with `findBy`. Perhaps both should be improved?
- `QueryData.updateMany` is stored as a single compound object, while insert batch data is still represented by flat `QueryData` fields. The two batch-mutation paths are conceptually similar but structurally inconsistent.

### CODE REVIEW

File: packages/pqb/src/query/basic-features/mutate/update.sql.ts [hunk 1]
```diff
@@ -259,3 +265,153 @@ const processValue = (
...
+      query as unknown as Query,
+      'updateMany cannot be combined with updateFrom',
+    );
+  }
+  if (q.join) {
+    throw new OrchidOrmInternalError(
+      query as unknown as Query,
+      'updateMany cannot be combined with join',
+    );
+  }
```
Comment (romeerez) [inline packages/pqb/src/query/basic-features/mutate/update.sql.ts]: The same checks are in `update.ts`, no need to duplicate them here.

File: packages/pqb/src/query/basic-features/mutate/update.sql.ts [hunk 2]
```diff
@@ -259,3 +265,153 @@ const processValue = (
...
+    const merged: RecordUnknown = {};
+    for (const item of q.hookUpdateSet) Object.assign(merged, item);
+    applySet(ctx, query, set, merged, perRowSkip, quotedAs);
+  }
+
+  // 5. Empty SET check
+  if (!set.length) {
+    throw new OrchidOrmInternalError(
+      query as unknown as Query,
+      'updateMany: no columns to set. Provide non-key columns in data or use .set()',
```
Comment (romeerez) [inline packages/pqb/src/query/basic-features/mutate/update.sql.ts]: it's better to be moved to `update.ts`: the sooner it validates the better.

File: packages/pqb/src/query/basic-features/mutate/update.sql.ts [hunk 3]
```diff
@@ -259,3 +265,153 @@ const processValue = (
...
+  }
+
+  // Build the UPDATE statement
+  ctx.sql.push(`UPDATE ${from}`);
+  if (`"${query.table || (q.from as string)}"` !== quotedAs) {
+    ctx.sql.push(quotedAs);
+  }
+  ctx.sql.push('SET', set.join(', '));
+  ctx.sql.push('FROM', valuesSql);
+  ctx.sql.push('WHERE', whereParts.join(' AND '));
```
Comment (romeerez) [inline packages/pqb/src/query/basic-features/mutate/update.sql.ts]: ### must have

`where` should be supported:
- you can specify `where({ ownerId: currentUser.id })` to disallow updating records that aren't yours
- soft delete filters by `deletedAt` field.

File: packages/pqb/src/query/basic-features/mutate/update.test.ts [hunk 4]
```diff
@@ -1283,3 +1285,380 @@ describe('update', () => {
...
+        'no_ts',
+        (t) => ({
+          id: t.identity().primaryKey(),
+          name: t.string(),
+        }),
+        undefined,
+        { schema: () => 'schema' },
+      );
+      expect(() => NoTimestamps.updateMany([{ id: 1 }]).toSQL()).toThrow(
+        'no columns to set',
```
Comment (romeerez) [inline packages/pqb/src/query/basic-features/mutate/update.test.ts]: And here is a second reason why the `set.length` check is in SQL: timestamps are mixed in after `toSQL` is called.

I checked how the regular `update` works: it turns `UPDATE` to `SELECT count(*)` when there is nothing to set.

It's expected that user's API accepts a `Partial` data for update, and technically an empty set is a valid payload. Later, user may find out that API is failing on empty sets, causing 500 because it wasn't handled, and then user has to handle that edge-case.

Selecting count is a way for ORM to handle that case so that user doesn't have to.

File: packages/pqb/src/query/basic-features/mutate/update.test.ts [hunk 5]
```diff
@@ -1283,3 +1285,380 @@ describe('update', () => {
...
+      );
+    });
+
+    it('should throw on duplicate keys', () => {
+      expect(() =>
+        User.updateMany([
+          { id: 1, name: 'a' },
+          { id: 1, name: 'b' },
+        ]),
+      ).toThrow('Duplicate key');
```
Comment (romeerez) [inline packages/pqb/src/query/basic-features/mutate/update.test.ts]: It's great to have you on my team, because I'd not bother with such edge cases :)

It makes sense, having "not found" instead would be confusing.

File: packages/pqb/src/query/basic-features/mutate/update.ts [hunk 6]
```diff
@@ -689,4 +872,159 @@ export class QueryUpdate {
...
+   *     { id: 2, name: 'Bob' },
+   *   ])
+   *   .set({ updatedBy: currentUser.id });
+   * ```
+   */
+  updateMany<
+    T extends PickQueryShapeResultReturnTypeSinglePrimaryKey & UpdateSelf,
+  >(
+    this: T,
+    data: (ShapePrimaryKeys<T['shape']> & Partial<T['inputType']>)[],
```
Comment (romeerez) [inline packages/pqb/src/query/basic-features/mutate/update.ts]: To be optimized, I'll handle.

Here two type mappings - `ShapePrimaryKeys` and `Partial` - are combined with `&`, it can be a single mapping.

**Review by romeerez — CHANGES_REQUESTED** [review]

**Comment by romeerez** [comment]
Thanks again and a lot for your efforts, please address the comments as much as you have time, and I'll help with the code on the weekends.

Most of all wish it was simpler. There are 3 loops over the same `data` in one of the `update` - I believe they can be combined, simplified. And need to try to reuse logic more with the existing update/insert. Even if `updateMany` only supports primitive values and expressions, a common function for processing values could be used. A common function for adjusting return type could be used. Need to look for cases like that, because it's quite a large addition, to be maintained. 

AI did a great job at writing tests, TDD FTW, refactoring is a pleasure when you have tests.

**Comment by romeerez** [comment]
Thank you for the hard efforts!

Don't make more changes, looks good overall, I'll review & refine in the following days.

**Comment by romeerez** [comment]
@IlyaSemenov merging with edits: https://github.com/romeerez/orchid-orm/pull/678

**Comment by romeerez** [comment]
You handled the types really good.
